src/infill/subDivCube.cpp

```diff
@@ -0,0 +1,279 @@
+
+    if (depth == 0) // lowest layer, no need for subdivision, exit.
+    {
+        return;
+    }
+    Point3 child_center;
+    coord_t radius = double(radius_multiplier * double(height[depth])) / 4.0 + radius_addition;
+
+    int child_nr = 0;
+    std::vector<Point3> rel_child_centers;
+    rel_child_centers.emplace_back(1, 1, 1); // top
+    rel_child_centers.emplace_back(-1, 1, 1); // top three
+    rel_child_centers.emplace_back(1, -1, 1);
+    rel_child_centers.emplace_back(1, 1, -1);
+    rel_child_centers.emplace_back(-1, -1, -1); // bottom
+    rel_child_centers.emplace_back(1, -1, -1); // bottom three
+    rel_child_centers.emplace_back(-1, 1, -1);
+    rel_child_centers.emplace_back(-1, -1, 1);
+    for (Point3 rel_child_center : rel_child_centers)
+    {
+        child_center = center + rotation_matrix.apply(rel_child_center * int32_t(side_length[depth] / 4));
+        if (isValidSubdivision(mesh, child_center, radius))
+        {
+            children[child_nr] = new SubDivCube(mesh, child_center, depth - 1);
+            child_nr++;
+        }
+    }
+}
+
+bool SubDivCube::isValidSubdivision(SliceMeshStorage& mesh, Point3& center, int64_t radius)
+{
+    int64_t distance2;
+    long int sphere_slice_radius2;//!< squared radius of bounding sphere slice on target layer
+    bool inside_somewhere = false;
+    bool outside_somewhere = false;
+    int inside;
+    double part_dist;//what percentage of the radius the target layer is away from the center along the z axis. 0 - 1
+    const long int layer_height = mesh.getSettingInMicrons("layer_height");
+    long int bottom_layer = (center.z - radius) / layer_height;
+    long int top_layer = (center.z + radius) / layer_height;
+    for (long int test_layer = bottom_layer; test_layer <= top_layer; test_layer += 3) // steps of three. Low-hanging speed gain.
+    {
+        part_dist = (double)(test_layer * layer_height - center.z) / radius;
+        sphere_slice_radius2 = radius * radius * (1.0 - (part_dist * part_dist));
+        Point loc(center.x, center.y);
+
+        inside = distanceFromPointToMesh(mesh, test_layer, loc, &distance2);
+        if (inside == 1)
+        {
+            inside_somewhere = true;
+        }
+        else
+        {
+            outside_somewhere = true;
+        }
+        if (outside_somewhere && inside_somewhere)
+        {
+            return true;
+        }
+        if ((inside != 2) && distance2 < sphere_slice_radius2)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+int SubDivCube::distanceFromPointToMesh(SliceMeshStorage& mesh, long int layer_nr, Point& location, int64_t* distance2)
+{
+    if (layer_nr < 0 || (unsigned long int)layer_nr >= mesh.layers.size()) //!< this layer is outside of valid range
+    {
+        return 2;
+    }
+    Polygons collide;
+    mesh.layers[layer_nr].getSecondOrInnermostWalls(collide);
+    Point centerpoint = location;
+    bool inside = collide.inside(centerpoint);
```

Comment:
If it's inside, you wouldn't need to compute the distance, right? I'd expect `distance2` would become 0 anyway.